src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,529 @@
+        public override async Task SendAsync(ArraySegment<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken)
+        {
+            ThrowIfNotConnected();
+
+            if (messageType != WebSocketMessageType.Binary &&
+                messageType != WebSocketMessageType.Text)
+            {
+                throw new ArgumentException($"Invalid message type: '{messageType}' specified in method 'SendAsync'.  Valid types are 'Binary' and 'Text'",
+                    nameof(messageType));
+            }
+
+            if (!endOfMessage && buffer.Array == null)
+                return;
+
+            if (!endOfMessage)
+            {
+                if (buffer.Array == null)
+                    return;
+                _writeBuffer = _writeBuffer ?? new MemoryStream();
+                _writeBuffer.Write(buffer.Array, buffer.Offset, buffer.Count);
+                return;
+            }
+            else if (_writeBuffer != null && buffer.Array != null)
+            {
+                _writeBuffer.Write(buffer.Array, buffer.Offset, buffer.Count);
+                if (!_writeBuffer.TryGetBuffer(out buffer))
+                    throw new WebSocketException(WebSocketError.NativeError);
+            }
+
+            var tcsSend = new TaskCompletionSource<bool>();
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully send or not.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            var writtenBuffer = _writeBuffer;
+            _writeBuffer = null;
+
+            using (cancellationToken.Register(() => tcsSend.TrySetCanceled()))
+            {
+                try
+                {
+                    if (messageType == WebSocketMessageType.Binary)
+                    {
+                        using (var uint8Buffer = Uint8Array.From(buffer))
+                        {
+                            _innerWebSocket!.Invoke("send", uint8Buffer);
+                            tcsSend.SetResult(true);
+                        }
+                    }
+                    else if (messageType == WebSocketMessageType.Text)
+                    {
+                        string strBuffer = buffer.Array == null ? string.Empty : Encoding.UTF8.GetString(buffer.Array, buffer.Offset, buffer.Count);
+                        _innerWebSocket!.Invoke("send", strBuffer);
+                        tcsSend.SetResult(true);
+                    }
+                }
+                catch (Exception excb)
+                {
+                    tcsSend.TrySetException(new WebSocketException(WebSocketError.NativeError, excb));
+                }
+                finally
+                {
+                    writtenBuffer?.Dispose();
+                }
+                await tcsSend.Task.ConfigureAwait(continueOnCapturedContext: true);
```

Comment:
In what situation will this complete asynchronously?  It seems like the whole operation here is actually synchronous.

Reply:
Addressed in recent commits